### PR TITLE
User log in

### PR DESCRIPTION
**Pull request by fottander**
https://www.pivotaltracker.com/story/show/141210483

We have used warden to login. And FactoryGirl to create a user. 

Ready for review!

@tochman @diraulo @AmberWilkie

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -20,7 +20,7 @@ def conversation
   protected
 
   def configure_permitted_parameters
-    devise_parameter_sanitizer.for(:sign_up) << :name
-    devise_parameter_sanitizer.for(:account_update) << :name
+    devise_parameter_sanitizer.permit(:sign_up, keys: [:name])
+    devise_parameter_sanitizer.permit(:account_update, keys: [:name])
   end
```
Comment (AmberWilkie) [inline app/controllers/application_controller.rb]: Do you need either of these things for your part of the program to work?

File: config/initializers/devise.rb [hunk 2]
```diff
@@ -262,4 +262,5 @@
   # When using OmniAuth, Devise cannot automatically set OmniAuth path,
   # so you need to do it manually. For the users scope, it would be:
   # config.omniauth_path_prefix = '/my_engine/users/auth'
+  config.secret_key = '21c76a4ec918d26fb886bf595430cb40570b2f2081334076a62a84365a23a93dcbe86c1fc2738b87adbe5cee955e7f06e13f2c0cef0e61987c6c3ef494355caf'
 end
```
Comment (AmberWilkie) [inline config/initializers/devise.rb]: This file should be excluded from version control.

File: config/secrets.yml [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+development:
+  secret_key_base: 2127f1349dcf8214a144dcd2cf2b401497f1ca59400f804530a2278a185d2e82592aab312293deb6968141292178598228da8cf9c9318baad5454794faadf8db
+
+test:
+  secret_key_base: 83a044f2a605f8e5183724edb8ce348b93c6623a7d677607ec98ac158df1f0987ebd35ad96c8641acb17e91706d388356aff9c2b4112f29b66e5d85ed021f446
+
+# Do not keep production secrets in the repository,
+# instead read values from the environment.
+production:
+  secret_key_base: <%= ENV["SECRET_KEY_BASE"] %>
```
Comment (AmberWilkie) [inline config/secrets.yml]: This file should be excluded from version control. Any time something is called "secrets", you should think about whether or not it belongs on the cloud.

File: features/log_in.feature [hunk 4]
```diff
@@ -0,0 +1,18 @@
+Feature: As a visitor.
+So that I can use the service.
+I would like to be able to go to login page and login on my account.
+
+Background:
+  Given the following users exist
+    | name  | email           | password     |
+    | felix | felix@telia.com | fisk1234          |
```
Comment (AmberWilkie) [inline features/log_in.feature]: Code formatting.

File: features/log_in.feature [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+
+Scenario: As I login
+  Given I am on the index page
+  And I click "Login" button
+  Then I should see "Log in"
+  When I fill in "user_email" with "felix@telia.com"
+  And I fill in "user_password" with "fisk1234"
```
Comment (AmberWilkie) [inline features/log_in.feature]: Can you not just do Email and Password?

File: features/step_definitions/log_in_steps.rb [hunk 6]
```diff
@@ -0,0 +1,23 @@
+And(/^I click "([^"]*)" button$/) do |button|
+  click_link_or_button button
+end
+
+When(/^I fill in "([^"]*)" with "([^"]*)"$/) do |arg1, arg2|
+  fill_in(arg1, with: arg2)
+  fill_in(arg1, with: arg2)
```
Comment (AmberWilkie) [inline features/step_definitions/log_in_steps.rb]: Why these duplicate lines?

File: features/step_definitions/log_in_steps.rb [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+end
+
+Then(/^I click "([^"]*)"$/) do |button|
+  click_link_or_button button
+end
+
+Given(/^I am logged in as "([^"]*)"$/) do |name|
+  user = User.find_by(name: name)
+  login_as(user, scope: :user)
```
Comment (AmberWilkie) [inline features/step_definitions/log_in_steps.rb]: Good. Glad to see you using the helper methods (remember helpers?) `login_as` is one written by the Devise library.

File: spec/factories/users.rb [hunk 8]
```diff
@@ -0,0 +1,7 @@
+FactoryGirl.define do
+  factory :user do
+    name 'felix'
+    user_email 'felix@telia.com'
+    user_password 'fisk1234'
```
Comment (AmberWilkie) [inline spec/factories/users.rb]: Now that you have removed `user_email` and `user_password` you should get rid of them from the factory.

**Review by AmberWilkie — COMMENTED** [review]
Ok fair enough. I want to make sure this works on another computer. Get someone to pull your code and run `rails db:migrate` and then `rake` to make sure it is working. If not, come get me.

File: config/secrets.yml (hunk 3 above)
Comment (AmberWilkie) [reply config/secrets.yml]: Scratch that - this file is fine.

File: features/log_in.feature [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+I would like to be able to go to login page and login on my account.
+
+Background:
+  Given the following users exist
+    | name  | email           | password     |
+    | felix | felix@telia.com | fisk1234     |
+
+Scenario: As I login
+  Given I am on the index page
+  And I click "Login" button
```
Comment (AmberWilkie) [inline features/log_in.feature]: Wait, why do you have two steps: `I click "Login" button` and `I click "Log in"`. These do the same thing, right?

File: features/step_definitions/log_in_steps.rb [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+  click_link_or_button button
+end
+
+When(/^I fill in "([^"]*)" with "([^"]*)"$/) do |arg1, arg2|
+  fill_in(arg1, with: arg2)
+end
+
+Then(/^I click "([^"]*)"$/) do |button|
+  click_link_or_button button
+end
```
Comment (AmberWilkie) [inline features/step_definitions/log_in_steps.rb]: Yes, they sure do. Keep one, delete the other.

**Review by AmberWilkie — COMMENTED** [review]
Ok one more thing before this is merged.

File: config/initializers/devise.rb [hunk 11]
```diff
@@ -262,4 +262,5 @@
   # When using OmniAuth, Devise cannot automatically set OmniAuth path,
   # so you need to do it manually. For the users scope, it would be:
   # config.omniauth_path_prefix = '/my_engine/users/auth'
+  config.secret_key = '21c76a4ec918d26fb886bf595430cb40570b2f2081334076a62a84365a23a93dcbe86c1fc2738b87adbe5cee955e7f06e13f2c0cef0e61987c6c3ef494355caf'
```
Comment (diraulo) [inline config/initializers/devise.rb]: You don't need to update this file, I solved the issue with this in my CI Fix PR. So revert the changes on this file

**Review by AmberWilkie — APPROVED** [review]

**Pull request merged**